test_sibling_enable_sameas started to FAIL on appveyor (Ubu) recently

spotted while reviewing https://github.com/datalad/datalad/pull/7318 .

looked like
```
FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
[{'action': 'get',
  'annexkey': 'SHA256E-s1--5feceb66ffc86f38d952786c6d696c79c2dbc239dd4e91b46729d73a27fb57e9',
  'error_message': 'rsync failed\nrsync failed\nrsync failed',
  'path': '/home/appveyor/DLTMP/datalad_temp_test_sibling_enable_sameast7tzb2z1/f0',
  'refds': '/home/appveyor/DLTMP/datalad_temp_test_sibling_enable_sameast7tzb2z1',
  'status': 'error',
  'type': 'file'}]

```

and apparently not for the first time but started recently:

```
$> git grep 'FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas'
08/pr/7317/d3a47bc/appveyor-9170-failed/1o7ovh42loqs1udv.txt:[00:25:10] [gw1] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas 
08/pr/7317/d3a47bc/appveyor-9170-failed/1o7ovh42loqs1udv.txt:[00:27:06] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
08/pr/7317/d3a47bc/appveyor-9170-failed/ulnb0prc4odiwjcl.txt:[00:23:38] [gw1] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas 
08/pr/7317/d3a47bc/appveyor-9170-failed/ulnb0prc4odiwjcl.txt:[00:25:24] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
08/push/maint/901c8f4/appveyor-9167-failed/ni3la8616gomqukn.txt:[00:24:30] [gw1] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas 
08/push/maint/901c8f4/appveyor-9167-failed/ni3la8616gomqukn.txt:[00:26:13] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
08/push/maint/901c8f4/appveyor-9167-failed/td81tnkiw3rfc5f6.txt:[00:22:52] [gw0] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas 
08/push/maint/901c8f4/appveyor-9167-failed/td81tnkiw3rfc5f6.txt:[00:24:51] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
08/push/maint/c1b079f/appveyor-9166-failed/hxrjorsunge47o1d.txt:[00:23:21] [gw1] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas 
08/push/maint/c1b079f/appveyor-9166-failed/hxrjorsunge47o1d.txt:[00:25:13] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
08/push/maint/c1b079f/appveyor-9166-failed/pe7etil04hh9e2ux.txt:[00:25:07] [gw1] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas 
08/push/maint/c1b079f/appveyor-9166-failed/pe7etil04hh9e2ux.txt:[00:27:07] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
09/pr/7317/03da886/appveyor-9178-failed/9v3v9awdts4ybclk.txt:[00:25:55] [gw1] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas 
09/pr/7317/03da886/appveyor-9178-failed/9v3v9awdts4ybclk.txt:[00:27:48] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
09/pr/7317/03da886/appveyor-9178-failed/n5emh379373mrc5m.txt:[00:23:53] [gw1] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas 
09/pr/7317/03da886/appveyor-9178-failed/n5emh379373mrc5m.txt:[00:25:38] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
09/pr/7317/43cb9f0/appveyor-9176-failed/9y2trs40gvv5edpw.txt:[00:23:15] [gw1] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas 
09/pr/7317/43cb9f0/appveyor-9176-failed/9y2trs40gvv5edpw.txt:[00:24:58] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
09/pr/7317/43cb9f0/appveyor-9176-failed/ob9l82pjtbmfk6w2.txt:[00:25:40] [gw1] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas 
09/pr/7317/43cb9f0/appveyor-9176-failed/ob9l82pjtbmfk6w2.txt:[00:27:43] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
09/pr/7317/ff72c14/appveyor-9173-failed/cc39wnjusvhy0og7.txt:[00:25:24] [gw1] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas 
09/pr/7317/ff72c14/appveyor-9173-failed/cc39wnjusvhy0og7.txt:[00:27:17] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
09/pr/7317/ff72c14/appveyor-9173-failed/difgngng03j2vqsq.txt:[00:23:41] [gw1] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas 
09/pr/7317/ff72c14/appveyor-9173-failed/difgngng03j2vqsq.txt:[00:25:27] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
09/pr/7318/e0acb7e/appveyor-9175-failed/1h871v79409t03a1.txt:[00:25:08] [gw1] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas 
09/pr/7318/e0acb7e/appveyor-9175-failed/1h871v79409t03a1.txt:[00:27:08] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
09/pr/7318/e0acb7e/appveyor-9175-failed/kskka21kr42q3kpn.txt:[00:22:47] [gw0] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas 
09/pr/7318/e0acb7e/appveyor-9175-failed/kskka21kr42q3kpn.txt:[00:24:50] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
09/push/maint/6c42103/appveyor-9171-failed/awlarnbxu4865miy.txt:[00:25:59] [gw1] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas 
09/push/maint/6c42103/appveyor-9171-failed/awlarnbxu4865miy.txt:[00:27:53] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
09/push/maint/6c42103/appveyor-9171-failed/r7wd7ljfhc6cabqw.txt:[00:23:37] [gw1] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas 
09/push/maint/6c42103/appveyor-9171-failed/r7wd7ljfhc6cabqw.txt:[00:25:27] FAILED ../datalad/distribution/tests/test_siblings.py::test_sibling_enable_sameas - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
```

--- a/changelog.d/pr-7342.md
+++ b/changelog.d/pr-7342.md
@@ -1,0 +1,3 @@
+### 🧪 Tests
+
+- Skip with_sameas_remote when rsync and annex are incompatible.  Fixes [#7320](https://github.com/datalad/datalad/issues/7320) via [PR #7342](https://github.com/datalad/datalad/pull/7342) (by [@bpoldrack](https://github.com/bpoldrack))

--- a/datalad/distribution/tests/test_siblings.py
+++ b/datalad/distribution/tests/test_siblings.py
@@ -397,7 +397,6 @@ def test_arg_missing(path=None, path2=None):
         annex_groupwanted='whatever')
 
 
-@pytest.mark.xfail(on_appveyor, reason="https://github.com/datalad/datalad/issues/7320")
 @with_sameas_remote
 @with_tempfile(mkdir=True)
 def test_sibling_enable_sameas(repo=None, clone_path=None):

--- a/datalad/support/external_versions.py
+++ b/datalad/support/external_versions.py
@@ -150,6 +150,33 @@ def _get_system_7z_version():
     lgr.debug("Could not determine version of 7z from stdout. %s", out)
 
 
+def get_rsync_version():
+
+    # This does intentionally not query the version of rsync itself, but
+    # that of the debian package it's installed with. Reason is in gh-7320,
+    # which results in the need to detect a patched-by-ubuntu version of rsync
+    # and therefore the package version, not the result of `rsync --version`.
+    from datalad.utils import on_linux, get_linux_distribution
+    if on_linux:
+        dist = get_linux_distribution()[0]
+        if dist in ['debian', 'ubuntu']:
+            out = _runner.run(['apt-cache', 'policy', 'rsync'],
+                              protocol=StdOutErrCapture)
+            for line in out['stdout'].splitlines():
+                parts = line.split()
+                if parts[0] == 'Installed:':
+                    ver = LooseVersion(parts[1])
+                    break
+        # If we have a debian package version, use this as rsync version.
+        # Otherwise report what `rsync --version` itself has to say.
+        if ver:
+            return ver
+    out = _runner.run(['rsync', '--version'], protocol=StdOutErrCapture)
+    # Expected first line:
+    # rsync version x protocol version y
+    return LooseVersion(out['stdout'].splitlines()[0].split()[2])
+
+
 class ExternalVersions(object):
     """Helper to figure out/use versions of the externals (modules, cmdline tools, etc).
 

--- a/datalad/tests/utils_pytest.py
+++ b/datalad/tests/utils_pytest.py
@@ -36,7 +36,10 @@ from datalad import utils
 from datalad.consts import ARCHIVES_TEMP_DIR
 from datalad.dochelpers import borrowkwargs
 
-from datalad.support.external_versions import external_versions
+from datalad.support.external_versions import (
+    external_versions,
+    get_rsync_version,
+)
 from datalad.support.keyring_ import MemoryKeyring
 from datalad.support.network import RI
 from datalad.support.vcr_ import *
@@ -1123,6 +1126,16 @@ def with_sameas_remote(func, autoenabled=False):
         if external_versions['cmd:system-ssh'] < '7.4' and \
            '8.20200522' < external_versions['cmd:annex'] < '8.20200720':
             pytest.skip("Test known to hang")
+
+        # A fix in rsync 3.2.4 broke compatibility with older annex versions.
+        # To make things a bit more complicated, ubuntu pulled that fix into
+        # their rsync package for 3.1.3-8.
+        # Issue: gh-7320
+        rsync_ver = get_rsync_version()
+        rsync_fixed = rsync_ver >= "3.1.3-8ubuntu" or rsync_ver >= "3.2.4"
+        if rsync_fixed and external_versions['cmd:annex'] < "10.20220504":
+            pytest.skip(f"rsync {rsync_ver} and git-annex "
+                        f"{external_versions['cmd:annex']} incompatible")
 
         sr_path, repo_path = args[-2:]
         fn_args = args[:-2]

